src/D2L.CodeStyle.Analyzers/Diagnostics.cs

```diff
@@ -514,5 +514,77 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "We require that event types be sealed."
 		);
+
+		public static readonly DiagnosticDescriptor MutableBaseType = new DiagnosticDescriptor(
+			id: "D2L0061",
+			title: "Base class for immutable type must be [Immutable] (or [ImmutableBaseClass])",
+			messageFormat: "{0}'s base class is {1} which is missing [Immutable] (or, more weakly, [ImmutableBaseClass]).",
+			category: "Immutability",
+			defaultSeverity: DiagnosticSeverity.Error,
+			isEnabledByDefault: true
+		);
+
+		public static readonly DiagnosticDescriptor ArraysAreMutable = new DiagnosticDescriptor(
+			id: "D2L0062",
+			title: "Arrays are mutable and thus can't be held by an immutable type.",
+			messageFormat: "{0}[]'s are mutable even if {0}'s are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<{0}>.",
```

Comment:
"Foo[]'s are mutable even if Foo's are immutable because array items can be substituted. Prefer an immutable container type such as ImmutableArray<Foo>."